prov/efa/src/efa_mr.c

```diff
@@ -547,6 +550,139 @@ static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr
 }
 #endif /* HAVE_SYNAPSEAI */
 
+/**
+ * @brief update the mr_map inside util_domain with a new memory registration.
+ *
+ * The mr_map is in util domain is a map between MR key and MR, and has all
+ * the active MR in it. This function add the information of a new MR into
+ * the mr_map.
+ *
+ * @param	efa_mr		the pointer to the efa_mr object, which has pointer to domain
+ * @param	mr_attr		the pointer to an fi_mr_attr object.
+ * @return	0 on success.
+ * 		negative libfabric error code on failure.
+ */
+#if HAVE_CUDA
+static
+int efa_mr_update_domain_mr_map(struct efa_mr *efa_mr, struct fi_mr_attr *mr_attr)
+{
+	struct fid_mr *existing_mr_fid;
+	struct efa_mr *existing_mr;
+	CUmemorytype cuda_memory_type;
+	int err;
+
+	mr_attr->requested_key = efa_mr->mr_fid.key;
+	ofi_genlock_lock(&efa_mr->domain->util_domain.lock);
+	err = ofi_mr_map_insert(&efa_mr->domain->util_domain.mr_map, mr_attr,
+				&efa_mr->mr_fid.key, &efa_mr->mr_fid);
+	ofi_genlock_unlock(&efa_mr->domain->util_domain.lock);
+
+	/* FI_ENOKEY can happen legitimately when the key was assigned a CUDA memory region, and the memory region has been freed
+	 * after the registration.
+	 *
+	 * This is because of the the combination of following 3 things:
+	 *  1) cuda memory uses memory cache.
+	 *  2) cuda memory cache's monitor is not monitoring call to cudaFree().
+	 *  3) kernel released the device part of memory registration when cudaFree() is called on a registered cuda memory.
+	 *
+	 * Therefore, when application call cudaFree() on a registered CUDA memory region, the device released the device
+	 * part of the memory region, make the key available for reuse. However, libfabric is not aware of the incident,
+	 * and kept the user space memory registration in its MR cache and MR map.
+	 *
+	 * The long term solution to handle this issue is to use a CUDA memory minitor that monitors call to cudaFree()
+	 * (cuda memhook monitor).
+	 * Before that is implemented, we deregister the staled user space memory registration and retry the insert.
+	 *
+	 * TODO: remove this branch once CUDA memhook monitor is implemented and used by efa.
+	 */
+	if (err == -FI_ENOKEY) {
+		existing_mr_fid = ofi_mr_map_get(&efa_mr->domain->util_domain.mr_map, efa_mr->mr_fid.key);
+		assert(existing_mr_fid);
+		existing_mr = container_of(existing_mr_fid, struct efa_mr, mr_fid);
+
+		if (existing_mr->peer.iface != FI_HMEM_CUDA) {
+			EFA_WARN(FI_LOG_DOMAIN, "key %ld already assigned to buffer: %p hmem_iface: %d length: %ld\n",
+				 existing_mr->mr_fid.key,
+				 existing_mr->ibv_mr->addr,
+				 existing_mr->peer.iface,
+				 existing_mr->ibv_mr->length);
+			return -FI_ENOKEY;
+		}
+
+		assert(existing_mr->peer.iface == FI_HMEM_CUDA);
+		err = ofi_cuPointerGetAttribute(&cuda_memory_type, CU_POINTER_ATTRIBUTE_MEMORY_TYPE,
+						(CUdeviceptr)existing_mr->ibv_mr->addr);
+		if (!err) {
+			EFA_WARN(FI_LOG_DOMAIN, "key %ld already assigned to cuda buffer: %p cuda_memor_type: %d length: %ld\n",
```

Comment:
Is this warning message correct? Should it be like `failed to get the cuda ptr attribute? ...`

Reply:
The message is correct. The branch is for `!err` (no error)

Reply:
So, when ofi_cuPointerGetAttribute returns success, there must an error in the mr cache, because it failed to re-use the existing mr region. Right?

Reply:
I don't understand this. If cuPointerGetAttribute passed, that doesn't mean anything. That addr could have been reallocated at some point.

Reply:
If cuPointerGetAttribute passed, it means there is an alive CUDA memory region that is registered with the key, which we cannot handle and has to return error.

The only situation we can handle is that there is a dead region in the cache, for which we can remove it from cache.

Reply:
>That addr could have been reallocated at some point.

In that case, shouldn't the cuda monitor have moved it to a dead region. so a `ofi_mr_cache_flush` could make it removed?

Reply:
> In that case, shouldn't the cuda monitor have moved it to the dead region list. so a ofi_mr_cache_flush could make it removed?

Either way, if a pointer in mr map is valid. There is no way the function is able to handle the error, and return error is the right approach.

I think I will re-arrange the code and adjust the comment to make the point more clear.

Reply:
If it's valid but the key doesn't match, that means that the key is invalid. I don't see why we can't kick it out of the cache.

Reply:
> If it's valid but the key doesn't match, that means that the key is invalid. I don't see why we can't kick it out of the cache.

I assume the `key` you refer to is the mr key.  But here the error is that the mr key is the same, but from a different cuda buffer.

Reply:
I think this is an erronenous situation, and we should return error

Reply:
I agree we need to return error here. But it is hard for me to understand how this could happen. EFA driver returns the same key from a different cuda buffer registration?

Reply:
I cannot think of a situation it will happen. But if it happened, we return error.